_posts/2018-05-25-ecs-calculator.md

```diff
@@ -0,0 +1,138 @@
+
+Let's assume your container normally uses 128 MBs of memory, but occasionally bursts to 256 MBs of memory for short periods of time. As an option, you can set a Memory Reservation of 128 MBs, and a Memory Limit of 300 MBs. This configuration would allow the container to only reserve 128 MBs of memory from the resources on the container instance, but also allows the container to consume more memory resources when needed.
+
+We can consider another scenario where you configure only Hard Memory Limit with no Memory Reservation. In this case the ECS container instance will deduct configured memory limit from memory resources of the container instance. We think that it is better to use exactly this option because by specifying only a Hard Memory Limit for your tasks you avoid running out of memory (ECS stops placing tasks on the instance, and docker kills any containers that try to go over the hard limit 😿), in other words, Hard Memory Limit exists for the optimal fit of containers in the instance.
+
+## How the data can help you to optimally scale and fit ECS containers in an EC2 instance?
+
+Based on the soft/hard limit of CPU/Memory reservation, the approach of solving our current issue is taken according to the following steps.
+* Obtaining the number of the Memory units required for your largest container.
+* Optimally choosing the EC2 instance type that can budget those Memory units.
+* Calculating CPU units needed for your largest container.
+
+The related examples are given below.
+
+## Examples
+
+Imagine that your largest container requires 332 memory units. If you choose `t2.micro` as the EC2 instance type which offers 993 memory units, you can fit 2 containers in the instance (993/332 = 2.99 containers). However, if you choose `t2.small` which offers 2001 memory units, the memory resource would be more efficiently used because 2001/332 = 6.02 containers and not using the leftover 0.02 containers would be less wasteful than doing this with the leftover 0.99 containers. In this way, your containers fit more efficiently in the `t2.small` instance than in the `t2.micro` instance. The next question is how can we calculate CPU for your container. The amount for CPU units you can easily calculate using this ratio:
+
+![ratio](/assets/posts/2018-05-25-ecs-calculator/ratio_cpu_ram.png)
+
+> So, (332 Mbs * 1024 CPU units) / 2001 Mbs = 169 CPU units.
+
+In another scenario, your largest container is provided with 1750 memory units. You can fit 1 container in the `t2.small` instance which reserves 2001 Memory units (2001/1750 = 1.14 containers). Else, the `t2.medium` instance which reserves 3952 Memory units makes it possible for 2 containers to fit in (3952/1750 = 2.25 containers). Then your containers will fit more efficiently in the `t2.small` instance than in the `t2.medium` instance with the same argument from the previous example applied. Let's calculate the CPU units then.
```

Comment:
It is not a good idea to host 1 container per instance, regardless of the ratio. You want to pick an instance that can host at least 4-5 containers and then try to optimize the ratio.

Reply:
Use the value of 900MB of RAM instead of 1750